# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/CodeGenerator.cpp
# Commit: cd7be783293a
# Full Hash: cd7be783293ac391410379cbd3c190c9ad761fa2
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2023-03-30 15:09:05
# Regressor Bug: 1825014
# File Overlap Count: 1
# Description:
#   Bug 1825014 part 1 - Assert GetFrameArgument index is in bounds in debug builds. r=iain
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D173951
# ==============================================================================

diff -r 8c35fddc4405 -r cd7be783293a js/src/jit/CodeGenerator.cpp
--- a/js/src/jit/CodeGenerator.cpp	Thu Mar 30 08:01:22 2023 +0000
+++ b/js/src/jit/CodeGenerator.cpp	Thu Mar 30 08:05:41 2023 +0000
@@ -13001,12 +13001,36 @@
   const LAllocation* index = lir->index();
   size_t argvOffset = JitFrameLayout::offsetOfActualArgs();
 
+  // This instruction is used to access actual arguments and formal arguments.
+  // The number of Values on the stack is |max(numFormals, numActuals)|, so we
+  // assert |index < numFormals || index < numActuals| in debug builds.
+  DebugOnly<size_t> numFormals = gen->outerInfo().script()->function()->nargs();
+
   if (index->isConstant()) {
     int32_t i = index->toConstant()->toInt32();
+#ifdef DEBUG
+    if (uint32_t(i) >= numFormals) {
+      Label ok;
+      Register argc = result.scratchReg();
+      masm.loadNumActualArgs(FramePointer, argc);
+      masm.branch32(Assembler::Above, argc, Imm32(i), &ok);
+      masm.assumeUnreachable("Invalid argument index");
+      masm.bind(&ok);
+    }
+#endif
     Address argPtr(FramePointer, sizeof(Value) * i + argvOffset);
     masm.loadValue(argPtr, result);
   } else {
     Register i = ToRegister(index);
+#ifdef DEBUG
+    Label ok;
+    Register argc = result.scratchReg();
+    masm.branch32(Assembler::Below, i, Imm32(numFormals), &ok);
+    masm.loadNumActualArgs(FramePointer, argc);
+    masm.branch32(Assembler::Above, argc, i, &ok);
+    masm.assumeUnreachable("Invalid argument index");
+    masm.bind(&ok);
+#endif
     BaseValueIndex argPtr(FramePointer, i, argvOffset);
     masm.loadValue(argPtr, result);
   }
